MachObfuscator/Mach/ObjC/ObjcSymbols+Impl.swift

```diff
@@ -0,0 +1,342 @@
+    private var class_ro: Arch.ClassData.RODataType {
+        return mach.getStruct(fromVmOffset: raw.data)
+    }
+}
+
+private struct ObjcCategoryImpl<Arch: ObjcArchitecture>: ObjcCategory, FromMach, CustomStringConvertible {
+    typealias Raw = ObjC.category_t<Arch.PointerType>
+    fileprivate let mach: Mach
+    fileprivate let offset: Int
+
+    init(mach: Mach, offset: Int) {
+        self.mach = mach
+        self.offset = offset
+    }
+
+    var name: MangledObjcClassNameInData {
+        return mach.getCString(fromVmOffset: raw.name)
+    }
+
+    var cls: ObjcClass? {
+        guard raw.cls != 0 else {
+            return nil
+        }
+        let offset = mach.fileOffsetIfExists(fromVmOffset: raw.cls)!
+        return ObjcClassImpl<Arch>(mach: mach, offset: offset)
+    }
+
+    var methods: [ObjcMethod] {
+        // methods_list is entsize_list_tt with flags=0x3
+        let list: [ObjcMethodImpl<Arch>] = mach.get_entsize_list_tt(fromVmOffset: raw.instanceMethods, flags: 0x03)
+        return list
+    }
+
+    var properties: [ObjcProperty] {
+        // property_list is entsize_list_tt with flags=0
+        let list: [ObjcPropertyImpl<Arch>] = mach.get_entsize_list_tt(fromVmOffset: raw.instanceProperties)
+        return list
+    }
+
+    var description: String {
+        return "Category \(cls?.name.value ?? "")+\(name) methods: \(methods) properties:\(properties)"
+    }
+}
+
+private struct ObjcProtocolImpl<Arch: ObjcArchitecture>: ObjcProtocol, FromMach, CustomStringConvertible {
+    typealias Raw = ObjC.protocol_t<Arch.PointerType>
+    fileprivate let mach: Mach
+    fileprivate let offset: Int
+
+    init(mach: Mach, offset: Int) {
+        self.mach = mach
+        self.offset = offset
+    }
+
+    var name: MangledObjcClassNameInData {
+        return mach.getCString(fromVmOffset: raw.name)
+    }
+
+    var methods: [ObjcMethod] {
+        // methods_list is entsize_list_tt with flags=0x3
+        let list: [ObjcMethodImpl<Arch>] = mach.get_entsize_list_tt(fromVmOffset: raw.instanceMethods, flags: 0x03)
+        return list
+    }
+
+    var properties: [ObjcProperty] {
+        // property_list is entsize_list_tt with flags=0
+        let list: [ObjcPropertyImpl<Arch>] = mach.get_entsize_list_tt(fromVmOffset: raw.instanceProperties)
+        return list
+    }
+
+    var description: String {
+        return "Protocol \(name) methods: \(methods) properties:\(properties)"
+    }
+}
+
+private class MachArchitecture<PointerType: UnsignedInteger, Me: ObjcArchitecture>: ObjcImage {
+    private let mach: Mach
+    init(mach: Mach) {
+        self.mach = mach
+    }
+
+    func getObjectsOffsetsFromList(section list: Mach.Section?) -> [Int] {
+        guard let objcList = list else {
+            return []
+        }
+        let objectVMAddresses: [PointerType] = mach.data.getStructs(fromRange: objcList.range.intRange)
+        let objectFileAddresses: [Int] = objectVMAddresses.map(mach.fileOffset(fromVmOffset:))
```

Comment:
Why using here this `fileOffset(fromVmOffset:)` while `fileOffsetIfExists` is used everywhere else? `fileOffset(fromVmOffset:)` returns the `vmOffset` in case it doesn't find the matching segment. Shouldn't app crash when such a case happens?

In case there is a reason – I think we could remove `fileOffset(fromVmOffset:)` method at all and change this line to e.g.:
```
let objectFileAddresses: [Int] = objectVMAddresses.map { mach.fileOffseIfExist(fromVmOffset:$0) ?? $0 }
```
because `fileOffset(fromVmOffset:)` doesn't say anything about this default behaviour, and this behaviour is used just in this one place.

Reply:
I was thinking of completely removing `fileOffset(fromVmOffset:)` (and renaming `fileOffsetIfExists` to `fileOffset`). The default value may have some sense only for 32-bit images (in this case file offset seems to be equal to vmoffset if MachOView is not cheating) but for 64-bit images it is always incorrect, because they are loaded ad 0x100000000 base

Reply:
Let's do this! :D We can think about default value when we detect any crash. So what do you think about renaming `fileOffsetIfExists` to `fileOffset`, making its return type non-optional and adding some `fatalError` inside the function?

Reply:
done